BUG: Multithreaded scaling bottlenecks on free-threaded Python

[This stackoverflow question](https://stackoverflow.com/questions/79851420/multithreading-becomes-much-slower-than-multiprocessing-in-free-threaded-python#79851420) seems to have identified two real scaling bottlenecks inside NumPy.

I build Python 3.15t from source using pyenv:

```
$ CFLAGS=-O2 pyenv install 3.15t-dev -k -g
...
$ pyenv global 3.15t-dev-debug
```

Then, in an up-to-date clone of the NumPy repo:

```
python -m pip install . -v -Cbuilddir=build -C'setup-args=-Dbuildtype=debugoptimized'
```

And finally run this script, which has to be named `mtmp.py`: https://gist.github.com/ngoldbaum/ff6428e2510e991247eba4c809dcc8ac. This is identical to what was posted on StackOverflow but with some code commented out.

On my M3 Macbook Pro, I get the following stdout running the script:

```
Inner loops 10, multithreading  time: 6.68 sec, result sum: 717434683.1879175
Inner loops 10, multiprocessing time: 4.86 sec, result sum: 717434683.1879175
```

If I run the script like so:

```
PYTHONPERFSUPPORT=1 samply record $(pyenv which python) mtmp.py
```

Then I get a profile output like this: https://share.firefox.dev/3Lde1Lm. I see two different scaling bottlenecks. First, there seem to be a mutex inside tracemalloc that has contention.

Second, the critical section I added to the array creation routines in https://github.com/numpy/numpy/pull/29394 is causing some contention in this case.

@ngoldbaum Could be my problem as I haven't read through the links above, but I got this error:
```
  In file included from ../numpy/_core/src/multiarray/scalartypes.c.src:16:
  In file included from ../numpy/_core/src/common/npy_pycompat.h:5:
  ../numpy/_core/src/common/pythoncapi-compat/pythoncapi_compat.h:2668:1: error: static declaration of 'PyUnstable_SetImmortal' follows non-static declaration
   2668 | PyUnstable_SetImmortal(PyObject *op)
        | ^
  /Users/charlie.liu/Desktop/hackumulation-v2/charliesWork/temp/pythonGitMain/include/python3.15t/cpython/object.h:497:17: note: previous declaration is here
    497 | PyAPI_FUNC(int) PyUnstable_SetImmortal(PyObject *op);
        |                 ^
  1 error generated.
  [146/329] Linking static target numpy/_core/libunique_hash.a
  [147/329] Compiling C object numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_loops.c.o
  FAILED: [code=1] numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_loops.c.o
  cc -Inumpy/_core/_multiarray_umath.cpython-315t-darwin.so.p -Inumpy/_core -I../numpy/_core -Inumpy/_core/include -I../numpy/_core/include -I../numpy/_core/src/common -I../numpy/_core/src/multiarray -I../numpy/_core/src/npymath -I../numpy/_core/src/umath -I../numpy/_core/src/highway -I/Users/charlie.liu/Desktop/hackumulation-v2/charliesWork/temp/pythonGitMain/include/python3.15t -I/private/var/folders/z5/4406xbpj6ld8jdbn676dh8980000gp/T/pip-req-build-1ysqao3l/.mesonpy-uicmxj22/meson_cpu -fvisibility=hidden -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c11 -O3 -fno-strict-aliasing -ftrapping-math -DNPY_HAVE_CLANG_FPSTRICT -DNPY_HAVE_NEON_VFPV4 -DNPY_HAVE_NEON_FP16 -DNPY_HAVE_NEON -DNPY_HAVE_ASIMD -F/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Accelerate.framework -idirafter/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Accelerate.framework/Headers -DACCELERATE_NEW_LAPACK -DHAVE_CBLAS '-DBLAS_SYMBOL_SUFFIX=$NEWLAPACK' -DNPY_INTERNAL_BUILD -DHAVE_NPY_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -MD -MQ numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_loops.c.o -MF numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_loops.c.o.d -o numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_loops.c.o -c numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/loops.c
  In file included from ../numpy/_core/src/umath/loops.c.src:19:
  In file included from ../numpy/_core/src/common/npy_pycompat.h:5:
  ../numpy/_core/src/common/pythoncapi-compat/pythoncapi_compat.h:2668:1: error: static declaration of 'PyUnstable_SetImmortal' follows non-static declaration
   2668 | PyUnstable_SetImmortal(PyObject *op)
        | ^
  /Users/charlie.liu/Desktop/hackumulation-v2/charliesWork/temp/pythonGitMain/include/python3.15t/cpython/object.h:497:17: note: previous declaration is here
    497 | PyAPI_FUNC(int) PyUnstable_SetImmortal(PyObject *op);
        |                 ^
  1 error generated.
  [148/329] Compiling Cython source /private/var/folders/z5/4406xbpj6ld8jdbn676dh8980000gp/T/pip-req-build-1ysqao3l/numpy/random/_common.pyx
  [149/329] Compiling C object numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/src_multiarray_abstractdtypes.c.o
  [150/329] Compiling C object numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_scalarmath.c.o
  FAILED: [code=1] numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_scalarmath.c.o
  cc -Inumpy/_core/_multiarray_umath.cpython-315t-darwin.so.p -Inumpy/_core -I../numpy/_core -Inumpy/_core/include -I../numpy/_core/include -I../numpy/_core/src/common -I../numpy/_core/src/multiarray -I../numpy/_core/src/npymath -I../numpy/_core/src/umath -I../numpy/_core/src/highway -I/Users/charlie.liu/Desktop/hackumulation-v2/charliesWork/temp/pythonGitMain/include/python3.15t -I/private/var/folders/z5/4406xbpj6ld8jdbn676dh8980000gp/T/pip-req-build-1ysqao3l/.mesonpy-uicmxj22/meson_cpu -fvisibility=hidden -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c11 -O3 -fno-strict-aliasing -ftrapping-math -DNPY_HAVE_CLANG_FPSTRICT -DNPY_HAVE_NEON_VFPV4 -DNPY_HAVE_NEON_FP16 -DNPY_HAVE_NEON -DNPY_HAVE_ASIMD -F/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Accelerate.framework -idirafter/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Accelerate.framework/Headers -DACCELERATE_NEW_LAPACK -DHAVE_CBLAS '-DBLAS_SYMBOL_SUFFIX=$NEWLAPACK' -DNPY_INTERNAL_BUILD -DHAVE_NPY_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -MD -MQ numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_scalarmath.c.o -MF numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_scalarmath.c.o.d -o numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_scalarmath.c.o -c numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/scalarmath.c
  In file included from ../numpy/_core/src/umath/scalarmath.c.src:27:
  In file included from ../numpy/_core/src/common/binop_override.h:8:
  In file included from ../numpy/_core/src/common/get_attr_string.h:5:
  In file included from ../numpy/_core/src/common/npy_pycompat.h:5:
  ../numpy/_core/src/common/pythoncapi-compat/pythoncapi_compat.h:2668:1: error: static declaration of 'PyUnstable_SetImmortal' follows non-static declaration
   2668 | PyUnstable_SetImmortal(PyObject *op)
        | ^
  /Users/charlie.liu/Desktop/hackumulation-v2/charliesWork/temp/pythonGitMain/include/python3.15t/cpython/object.h:497:17: note: previous declaration is here
    497 | PyAPI_FUNC(int) PyUnstable_SetImmortal(PyObject *op);
        |                 ^
  1 error generated.
  [151/329] Compiling C object numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_nditer_templ.c.o
  [152/329] Compiling Cython source numpy/random/mtrand.pyx
  [153/329] Compiling C++ object numpy/_core/libhighway_qsort.dispatch.h_ASIMD.a.p/src_npysort_highway_qsort.dispatch.cpp.o
  [154/329] Compiling C object numpy/_core/libloops_autovec.dispatch.h_baseline.a.p/meson-generated_loops_autovec.dispatch.c.o
  [155/329] Compiling C object numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_matmul.c.o
  [156/329] Compiling Cython source numpy/random/_generator.pyx
  [157/329] Compiling C object numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_einsum_sumprod.c.o
  [158/329] Compiling C object numpy/_core/_multiarray_umath.cpython-315t-darwin.so.p/meson-generated_lowlevel_strided_loops.c.o
  ninja: build stopped: subcommand failed.
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> No available output.
```
my python build is `Python 3.15.0a6+` and I was installing via 
`bin/python3.15t -m pip install -v --no-cache-dir  git+https://github.com/numpy/numpy.git@refs/pull/30826/head `

--- a/doc/release/upcoming_changes/30846.compatibility.rst
+++ b/doc/release/upcoming_changes/30846.compatibility.rst
@@ -1,4 +1,4 @@
 Default memory allocator change
-===============================
+-------------------------------
 NumPy now uses ``PyMem_RawMalloc`` and ``PyMem_RawFree`` as the default memory allocator,
 instead of system's ``malloc`` and ``free`` directly.

--- a/doc/release/upcoming_changes/30846.compatibility.rst
+++ b/doc/release/upcoming_changes/30846.compatibility.rst
@@ -1,0 +1,4 @@
+Default memory allocator change
+===============================
+NumPy now uses ``PyMem_RawMalloc`` and ``PyMem_RawFree`` as the default memory allocator,
+instead of system's ``malloc`` and ``free`` directly.

--- a/doc/release/upcoming_changes/30846.performance.rst
+++ b/doc/release/upcoming_changes/30846.performance.rst
@@ -1,0 +1,18 @@
+Improved scaling of ufuncs on free-threading
+============================================
+
+NumPy's ufuncs now scale significantly better on free-threading builds
+of CPython due to the following optimizations:
+
+* **Lock-free dispatch table:** The ufuncs dispatch table is now
+  implemented as a lock-free concurrent hash map, allowing multiple threads
+  to call ufuncs without contention.
+
+* **Immortal shared objects:** Certain shared objects, such as global memory
+  handlers, have been made immortal. This effectively reduces reference
+  counting contention across threads.
+
+* **Optimized memory allocation:** NumPy now utilizes ``PyMem_RawMalloc`` and
+  ``PyMem_RawFree`` for memory allocation. On Python 3.15 and newer, this
+  leverages ``mimalloc`` and significantly reduces memory allocation overhead
+  in multi-threaded workloads.

--- a/doc/release/upcoming_changes/30846.performance.rst
+++ b/doc/release/upcoming_changes/30846.performance.rst
@@ -1,5 +1,5 @@
 Improved scaling of ufuncs on free-threading
-============================================
+--------------------------------------------
 
 NumPy's ufuncs now scale significantly better on free-threading builds
 of CPython due to the following optimizations:

--- a/numpy/_core/src/multiarray/alloc.cpp
+++ b/numpy/_core/src/multiarray/alloc.cpp
@@ -252,7 +252,7 @@ npy_alloc_cache_dim(npy_uintp sz)
         sz = 2;
     }
     return _npy_alloc_cache(sz, sizeof(npy_intp), NBUCKETS_DIM, dimcache,
-                            &PyArray_malloc);
+                            &PyMem_RawMalloc);
 }
 
 NPY_NO_EXPORT void

--- a/numpy/_core/src/multiarray/alloc.cpp
+++ b/numpy/_core/src/multiarray/alloc.cpp
@@ -75,12 +75,12 @@ typedef struct cache_destructor {
     ~cache_destructor() {
         for (npy_uint i = 0; i < NBUCKETS; ++i) {
             while (datacache[i].available > 0) {
-                free(datacache[i].ptrs[--datacache[i].available]);
+                PyMem_RawFree(datacache[i].ptrs[--datacache[i].available]);
             }
         }
         for (npy_uint i = 0; i < NBUCKETS_DIM; ++i) {
             while (dimcache[i].available > 0) {
-                PyArray_free(dimcache[i].ptrs[--dimcache[i].available]);
+                PyMem_RawFree(dimcache[i].ptrs[--dimcache[i].available]);
             }
         }
     }
@@ -217,7 +217,6 @@ npy_alloc_cache_zero(size_t nmemb, size_t size)
 {
     void * p;
     size_t sz = nmemb * size;
-    NPY_BEGIN_THREADS_DEF;
     if (sz < NBUCKETS) {
         p = _npy_alloc_cache(sz, 1, NBUCKETS, datacache, &PyDataMem_NEW);
         if (p) {
@@ -225,9 +224,7 @@ npy_alloc_cache_zero(size_t nmemb, size_t size)
         }
         return p;
     }
-    NPY_BEGIN_THREADS;
     p = PyDataMem_NEW_ZEROED(nmemb, size);
-    NPY_END_THREADS;
     if (p) {
         indicate_hugepages(p, sz);
     }
@@ -296,10 +293,10 @@ PyDataMem_NEW(size_t size)
     void *result;
 
     assert(size != 0);
-    result = malloc(size);
+    result = PyMem_RawMalloc(size);
     int ret = PyTraceMalloc_Track(NPY_TRACE_DOMAIN, (npy_uintp)result, size);
     if (ret == -1) {
-        free(result);
+        PyMem_RawFree(result);
         return NULL;
     }
     return result;
@@ -313,10 +310,10 @@ PyDataMem_NEW_ZEROED(size_t nmemb, size_t size)
 {
     void *result;
 
-    result = calloc(nmemb, size);
+    result = PyMem_RawCalloc(nmemb, size);
     int ret = PyTraceMalloc_Track(NPY_TRACE_DOMAIN, (npy_uintp)result, nmemb * size);
     if (ret == -1) {
-        free(result);
+        PyMem_RawFree(result);
         return NULL;
     }
     return result;
@@ -329,7 +326,7 @@ NPY_NO_EXPORT void
 PyDataMem_FREE(void *ptr)
 {
     PyTraceMalloc_Untrack(NPY_TRACE_DOMAIN, (npy_uintp)ptr);
-    free(ptr);
+    PyMem_RawFree(ptr);
 }
 
 /*NUMPY_API
@@ -342,10 +339,10 @@ PyDataMem_RENEW(void *ptr, size_t size)
 
     assert(size != 0);
     PyTraceMalloc_Untrack(NPY_TRACE_DOMAIN, (npy_uintp)ptr);
-    result = realloc(ptr, size);
+    result = PyMem_RawRealloc(ptr, size);
     int ret = PyTraceMalloc_Track(NPY_TRACE_DOMAIN, (npy_uintp)result, size);
     if (ret == -1) {
-        free(result);
+        PyMem_RawFree(result);
         return NULL;
     }
     return result;
@@ -357,7 +354,7 @@ PyDataMem_RENEW(void *ptr, size_t size)
 static inline void *
 default_malloc(void *NPY_UNUSED(ctx), size_t size)
 {
-    return _npy_alloc_cache(size, 1, NBUCKETS, datacache, &malloc);
+    return _npy_alloc_cache(size, 1, NBUCKETS, datacache, &PyMem_RawMalloc);
 }
 
 // The default data mem allocator calloc routine does not make use of a ctx.
@@ -368,20 +365,17 @@ default_calloc(void *NPY_UNUSED(ctx), size_t nelem, size_t elsize)
 {
     void * p;
     size_t sz = nelem * elsize;
-    NPY_BEGIN_THREADS_DEF;
     if (sz < NBUCKETS) {
-        p = _npy_alloc_cache(sz, 1, NBUCKETS, datacache, &malloc);
+        p = _npy_alloc_cache(sz, 1, NBUCKETS, datacache, &PyMem_RawMalloc);
         if (p) {
             memset(p, 0, sz);
         }
         return p;
     }
-    NPY_BEGIN_THREADS;
-    p = calloc(nelem, elsize);
+    p = PyMem_RawCalloc(nelem, elsize);
     if (p) {
         indicate_hugepages(p, sz);
     }
-    NPY_END_THREADS;
     return p;
 }
 
@@ -391,7 +385,7 @@ default_calloc(void *NPY_UNUSED(ctx), size_t nelem, size_t elsize)
 static inline void *
 default_realloc(void *NPY_UNUSED(ctx), void *ptr, size_t new_size)
 {
-    return realloc(ptr, new_size);
+    return PyMem_RawRealloc(ptr, new_size);
 }
 
 // The default data mem allocator free routine does not make use of a ctx.
@@ -400,7 +394,7 @@ default_realloc(void *NPY_UNUSED(ctx), void *ptr, size_t new_size)
 static inline void
 default_free(void *NPY_UNUSED(ctx), void *ptr, size_t size)
 {
-    _npy_free_cache(ptr, size, NBUCKETS, datacache, &free);
+    _npy_free_cache(ptr, size, NBUCKETS, datacache, &PyMem_RawFree);
 }
 
 /* Memory handler global default */

